### PR TITLE
(chore) Bloom store: Rewrite block downloading/caching in bloom store fetcher

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

1. The `BloomStore` interface gets a new function `FetchBlocks()`, which accepts `[]BlockRef` and returns `[]BlockDirectory`. (see commit message of https://github.com/grafana/loki/pull/11857/commits/3cad5e8c70d640bee29c18db7e9a99a66b4f7d79)

1. The `{Meta,Block}Client` interfaces are unified: Both clients define the same set of operations:

       Get{Meta,Block}
       Get{Meta,Block}s
       Put{Meta,Block}
       Delete{Meta,Block}s

1. The `blockDownloader` is replaced by a simple queue powered by a channel. The queue is implemented using generics, so it can be re-used for different request/response types.

1. Code related to "block caching" moved into a separate file `cache.go` and unused code is removed.

### CODE REVIEW

File: pkg/storage/stores/shipper/bloomshipper/client.go [hunk 1]
```diff
@@ -149,68 +147,72 @@ func (b *BloomClient) PutMeta(ctx context.Context, meta Meta) error {
...
-		_, err = block.Data.Seek(0, 0)
-		if err != nil {
-			return fmt.Errorf("error uploading block file: %w", err)
-		}
+	return NewBlockDirectory(ref, path, b.logger), nil
+}
 
-		err = b.client.PutObject(ctx, key, block.Data)
+func (b *BloomClient) GetBlocks(ctx context.Context, refs []BlockRef) ([]BlockDirectory, error) {
+	// TODO(chaudum): Integrate download queue
```
Comment (salvacorts) [inline pkg/storage/stores/shipper/bloomshipper/client.go:181]: According to the PR desc:
> The blockDownloader is replaced by a simple queue powered by a channel.

Is this TODO outdated? I don't see any queue mechanism at this level, can't we reuse what's implemented in `FetchBlocksWithQueue`?

File: pkg/storage/stores/shipper/bloomshipper/fetcher.go [hunk 2]
```diff
@@ -108,71 +125,231 @@ func (f *Fetcher) writeBackMetas(ctx context.Context, metas []Meta) error {
...
+	keys := make([]string, 0, len(refs))
+	for _, ref := range refs {
+		keys = append(keys, f.client.Block(ref).Addr())
+	}
+	cacheHits, cacheBufs, _, err := f.blocksCache.Fetch(ctx, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []BlockDirectory
```
Comment (salvacorts) [inline pkg/storage/stores/shipper/bloomshipper/fetcher.go:199]: We can preallocate this with len(refs)

File: pkg/storage/stores/shipper/bloomshipper/client.go (hunk 1 above)
Comment (chaudum) [reply pkg/storage/stores/shipper/bloomshipper/client.go:181]: Yes, the TODO is outdated and should be removed.

File: pkg/storage/stores/shipper/bloomshipper/cache_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+			cached := BlockDirectory{
+				Path:                        extractedBlockDirectory,
+				removeDirectoryTimeout:      500 * time.Millisecond,
+				activeQueriersCheckInterval: 50 * time.Millisecond,
+				logger:                      log.NewLogfmtLogger(os.Stderr),
+				activeQueriers:              atomic.NewInt32(1),
+			}
+			cached.removeDirectoryAsync()
+			//ensure directory exists
+			require.Never(t, func() bool {
```
[the remark this answers is not in the document]
Comment (chaudum) [reply pkg/storage/stores/shipper/bloomshipper/cache_test.go:46]: Sure. I did not change the test case after moving it into the separate file.

File: pkg/storage/stores/shipper/bloomshipper/client.go [hunk 4]
```diff
@@ -240,66 +242,30 @@ func (b *BloomClient) GetMetas(ctx context.Context, refs []MetaRef) ([]Meta, err
...
-	step := int64(periodConfig.IndexTables.Period.Seconds())
-	lower := interval.Start.Unix() / step
-	upper := interval.End.Unix() / step
-	tables := make([]string, 0, 1+upper-lower)
-	for i := lower; i <= upper; i++ {
-		tables = append(tables, fmt.Sprintf("%s%d", periodConfig.IndexTables.Prefix, i))
+func findPeriod(configs []config.PeriodConfig, ts model.Time) (config.DayTime, error) {
+	for i := len(configs) - 1; i >= 0; i-- {
+		periodConfig := configs[i]
+		if periodConfig.From.Before(ts) || periodConfig.From.Equal(ts) {
```
Comment (owen-d) [inline pkg/storage/stores/shipper/bloomshipper/client.go:266]: nit
```suggestion
		if !periodConfig.From.After(ts) {
```

**Review by owen-d — APPROVED** [review]

**Pull request merged**